lib/plugins/openaps.js

```diff
@@ -0,0 +1,248 @@
+'use strict';
+
+var _ = require('lodash');
+var moment = require('moment');
+var times = require('../times');
+var levels = require('../levels');
+var timeago = require('./timeago')();
+
+function init() {
+
+  var openaps = {
+    name: 'openaps'
+    , label: 'OpenAPS'
+    , pluginType: 'pill-status'
+  };
+
+  openaps.getPrefs = function getPrefs(sbx) {
+    return {
+      warn: sbx.extendedSettings.warn ? sbx.extendedSettings.warn : 30,
+      urgent: sbx.extendedSettings.urgent ? sbx.extendedSettings.urgent : 60,
+      enableAlerts: sbx.extendedSettings.enableAlerts
+    };
+  };
+
+  openaps.setProperties = function setProperties (sbx) {
+    sbx.offerProperty('openaps', function setopenaps ( ) {
+      var openapsStatus = _.findLast(sbx.data.devicestatus, function (status) {
+        return sbx.entryMills(status) <= sbx.time && ('openaps' in status);
+      });
+
+      var loopStatus = checkLoopStatus(openapsStatus, sbx);
+      openapsStatus = openapsStatus || {};
+      openapsStatus.status = loopStatus;
+
+      return openapsStatus;
+    });
+  };
+
+  openaps.getEventTypes = function getEventTypes ( ) {
+    return [
+      {
+        val: 'OpenAPS Offline'
+        , name: 'OpenAPS Offline'
+        , bg: false, insulin: false, carbs: false, prebolus: false, duration: true, percent: false, absolute: false, profile: false, split: false
+      }
+    ];
+  };
+
+  openaps.checkNotifications = function checkNotifications(sbx) {
+    var prefs = openaps.getPrefs(sbx);
+
+    if (!prefs.enableAlerts) { return; }
+
+    var offlineMarker = _.findLast(sbx.data.treatments, function match (treatment) {
+      var eventTime = sbx.entryMills(treatment);
+      var eventEnd = treatment.duration ? eventTime + times.mins(treatment.duration).msecs : eventTime;
+      return eventTime <= sbx.time && treatment.eventType === 'OpenAPS Offline' && eventEnd >= sbx.time
+    });
+
+    if (offlineMarker) {
+      console.info('OpenAPS known offline, not checking for alerts');
+      return;
+    }
+
+    var prop = sbx.properties.openaps;
+    var ops = prepOps(prop, sbx);
+
+    if (!ops.lastLoopTime) {
+      console.info('OpenAPS hasn\'t reported a loop yet');
+      return;
+    }
+
+    var now = moment();
+    var level = levels.NONE;
+    if (ops.lastLoopTime.add(prefs.urgent, 'minutes').isBefore(now)) {
+      level = levels.URGENT;
+    } else if (ops.lastLoopTime.add(prefs.warn, 'minutes').isBefore(now)) {
+      level = levels.WARN;
+    }
+
+    if (level >= levels.WARN) {
+      sbx.notifications.requestNotify({
+        level: level
+        , title: 'OpenAPS isn\'t looping'
+        , message: 'Last Loop: ' + formatAgo(ops.lastLoopTime, now.valueOf()) + ' (' + ops.lastLoopTime.format('LT') + ')'
+        , pushoverSound: 'echo'
+        , plugin: openaps
+        , debug: ops
+      });
+    }
+  };
+
+  openaps.updateVisualisation = function updateVisualisation (sbx) {
```

Comment:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [This function's cyclomatic complexity is too high. (10)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1503885472/issues/source?bid=2452379&fileBranchId=2989096#l93)